File: crates/converge-core/src/codec.rs
//! Canonical binary form of a [`Document`].
//!
//! The same bytes serve two purposes: they are the input to the convergence
//! hash, and (with a small header) they are the snapshot format. A snapshot
//! therefore round-trips to exactly the bytes that were hashed. All integers
//! are little-endian and fixed width; objects are ordered by id and props by
//! key, so the encoding is a pure function of the logical state.

use std::collections::BTreeMap;

use crate::document::{Document, ObjectState, Register};
use crate::hlc::{Hlc, Stamp};
use crate::ids::{OpId, ReplicaId};
use crate::value::{ObjectKind, Value};
use crate::Hash;

const SNAPSHOT_MAGIC: &[u8; 4] = b"CVGS";
const SNAPSHOT_VERSION: u8 = 1;

#[derive(thiserror::Error, Debug, PartialEq, Eq)]
pub enum DecodeError {
    #[error("unexpected end of input")]
    Eof,
    #[error("bad magic or version")]
    BadHeader,
    #[error("invalid tag {0}")]
    BadTag(u8),
    #[error("invalid utf-8")]
    Utf8,
    #[error("trailing bytes")]
    Trailing,
}

struct Writer(Vec<u8>);

impl Writer {
    fn u8(&mut self, v: u8) {
        self.0.push(v);
    }
    fn u16(&mut self, v: u16) {
        self.0.extend_from_slice(&v.to_le_bytes());
    }
    fn u32(&mut self, v: u32) {
        self.0.extend_from_slice(&v.to_le_bytes());
    }
    fn u64(&mut self, v: u64) {
        self.0.extend_from_slice(&v.to_le_bytes());
    }
    fn i64(&mut self, v: i64) {
        self.0.extend_from_slice(&v.to_le_bytes());
    }
    fn bytes(&mut self, b: &[u8]) {
        self.u32(b.len() as u32);
        self.0.extend_from_slice(b);
    }
    fn stamp(&mut self, s: &Stamp) {
        self.u64(s.hlc.wall_ms);
        self.u16(s.hlc.logical);
        self.u64(s.replica.0);
    }
    fn value(&mut self, v: &Value) {
        self.u8(v.tag());
        match v {
            Value::Null => {}
            Value::Bool(b) => self.u8(*b as u8),
            Value::I64(i) => self.i64(*i),
            Value::F64(f) => self.u64(f.to_bits()),
            Value::Str(s) | Value::FracIndex(s) => self.bytes(s.as_bytes()),
            Value::Color(c) => self.u32(*c),
            Value::ObjRef(id) => {
                self.u64(id.replica.0);
                self.u64(id.counter);
            }
        }
    }
}

struct Reader<'a> {
    buf: &'a [u8],
    pos: usize,
}

impl<'a> Reader<'a> {
    fn take(&mut self, n: usize) -> Result<&'a [u8], DecodeError> {
        if self.pos + n > self.buf.len() {
            return Err(DecodeError::Eof);
        }
        let s = &self.buf[self.pos..self.pos + n];
        self.pos += n;
        Ok(s)
    }
    fn u8(&mut self) -> Result<u8, DecodeError> {
        Ok(self.take(1)?[0])
    }
    fn u16(&mut self) -> Result<u16, DecodeError> {
        Ok(u16::from_le_bytes(self.take(2)?.try_into().unwrap()))
    }
    fn u32(&mut self) -> Result<u32, DecodeError> {
        Ok(u32::from_le_bytes(self.take(4)?.try_into().unwrap()))
    }
    fn u64(&mut self) -> Result<u64, DecodeError> {
        Ok(u64::from_le_bytes(self.take(8)?.try_into().unwrap()))
    }
    fn i64(&mut self) -> Result<i64, DecodeError> {
        Ok(i64::from_le_bytes(self.take(8)?.try_into().unwrap()))
    }
    fn string(&mut self) -> Result<String, DecodeError> {
        let n = self.u32()? as usize;
        let b = self.take(n)?;
        String::from_utf8(b.to_vec()).map_err(|_| DecodeError::Utf8)
    }
    fn stamp(&mut self) -> Result<Stamp, DecodeError> {
        let wall = self.u64()?;
        let logical = self.u16()?;
        let replica = self.u64()?;
        Ok(Stamp::new(Hlc::new(wall, logical), ReplicaId(replica)))
    }
    fn value(&mut self) -> Result<Value, DecodeError> {
        let tag = self.u8()?;
        Ok(match tag {
            0 => Value::Null,
            1 => Value::Bool(self.u8()? != 0),
            2 => Value::I64(self.i64()?),
            3 => Value::F64(f64::from_bits(self.u64()?)),
            4 => Value::Str(self.string()?),
            5 => Value::Color(self.u32()?),
            6 => Value::FracIndex(self.string()?),
            7 => {
                let r = self.u64()?;
                let c = self.u64()?;
                Value::ObjRef(OpId::new(r, c))
            }
            t => return Err(DecodeError::BadTag(t)),
        })
    }
}

/// Canonical bytes of the full state (ghosts and tombstones included).
pub fn canonical_bytes(doc: &Document) -> Vec<u8> {
    let mut w = Writer(Vec::new());
    w.u64(doc.len() as u64);
    for (id, obj) in doc.objects() {
        w.u64(id.replica.0);
        w.u64(id.counter);
        w.u8(obj.kind.map(ObjectKind::tag).unwrap_or(0));
        w.u8(obj.deleted.value as u8);
        w.stamp(&obj.deleted.at);
        w.u32(obj.props.len() as u32);
        for (k, reg) in &obj.props {
            w.bytes(k.as_bytes());
            w.value(&reg.value);
            w.stamp(&reg.at);
        }
    }
    w.0
}

/// BLAKE3 over the canonical bytes — the convergence oracle.
pub fn hash(doc: &Document) -> Hash {
    *blake3::hash(&canonical_bytes(doc)).as_bytes()
}

pub fn hash_hex(doc: &Document) -> String {
    hex(&hash(doc))
}

pub fn hex(bytes: &[u8]) -> String {
    bytes.iter().map(|b| format!("{b:02x}")).collect()
}

/// Snapshot = header + canonical bytes.
pub fn encode_snapshot(doc: &Document) -> Vec<u8> {
    let mut out = Vec::with_capacity(5);
    out.extend_from_slice(SNAPSHOT_MAGIC);
    out.push(SNAPSHOT_VERSION);
    out.extend(canonical_bytes(doc));
    out
}

pub fn decode_snapshot(bytes: &[u8]) -> Result<Document, DecodeError> {
    if bytes.len() < 5 || &bytes[..4] != SNAPSHOT_MAGIC || bytes[4] != SNAPSHOT_VERSION {
        return Err(DecodeError::BadHeader);
    }
    let mut r = Reader { buf: bytes, pos: 5 };
    let mut doc = Document::new();
    let n = r.u64()?;
    for _ in 0..n {
        let replica = r.u64()?;
        let counter = r.u64()?;
        let kind_tag = r.u8()?;
        let kind = if kind_tag == 0 {
            None
        } else {
            Some(ObjectKind::from_tag(kind_tag).ok_or(DecodeError::BadTag(kind_tag))?)
        };
        let deleted_v = r.u8()? != 0;
        let deleted_at = r.stamp()?;
        let np = r.u32()?;
        let mut props = BTreeMap::new();
        for _ in 0..np {
            let key = r.string()?;
            let value = r.value()?;
            let at = r.stamp()?;
            props.insert(key, Register::new(value, at));
        }
        doc.insert_raw(
            OpId::new(replica, counter),
            ObjectState {
                kind,
                deleted: Register::new(deleted_v, deleted_at),
                props,
            },
        );
    }
    if r.pos != bytes.len() {
        return Err(DecodeError::Trailing);
    }
    Ok(doc)
}

File: web/packages/engine/src/codec.ts
/**
 * Canonical bytes, hash and snapshot format — byte-identical to the Rust
 * `codec` module. Little-endian, fixed-width integers.
 */
import { blake3 } from "@noble/hashes/blake3";
import { Document, ObjectState, type Register } from "./document.js";
import type { Hlc, Stamp } from "./hlc.js";
import { opId, type OpId } from "./ids.js";
import { KIND_TAG, TAG_KIND, VALUE_TAG, type Value } from "./value.js";

const SNAPSHOT_MAGIC = [0x43, 0x56, 0x47, 0x53]; // "CVGS"
const SNAPSHOT_VERSION = 1;
const utf8 = new TextEncoder();
const utf8d = new TextDecoder("utf-8", { fatal: true });

class Writer {
  private buf = new Uint8Array(1024);
  private view = new DataView(this.buf.buffer);
  private pos = 0;

  private ensure(n: number): void {
    if (this.pos + n <= this.buf.length) return;
    let cap = this.buf.length * 2;
    while (cap < this.pos + n) cap *= 2;
    const nb = new Uint8Array(cap);
    nb.set(this.buf.subarray(0, this.pos));
    this.buf = nb;
    this.view = new DataView(nb.buffer);
  }
  u8(v: number): void {
    this.ensure(1);
    this.view.setUint8(this.pos, v);
    this.pos += 1;
  }
  u16(v: number): void {
    this.ensure(2);
    this.view.setUint16(this.pos, v, true);
    this.pos += 2;
  }
  u32(v: number): void {
    this.ensure(4);
    this.view.setUint32(this.pos, v >>> 0, true);
    this.pos += 4;
  }
  u64(v: bigint): void {
    this.ensure(8);
    this.view.setBigUint64(this.pos, BigInt.asUintN(64, v), true);
    this.pos += 8;
  }
  i64(v: bigint): void {
    this.ensure(8);
    this.view.setBigInt64(this.pos, v, true);
    this.pos += 8;
  }
  f64(v: number): void {
    this.ensure(8);
    this.view.setFloat64(this.pos, v, true);
    this.pos += 8;
  }
  raw(b: Uint8Array): void {
    this.ensure(b.length);
    this.buf.set(b, this.pos);
    this.pos += b.length;
  }
  bytes(b: Uint8Array): void {
    this.u32(b.length);
    this.raw(b);
  }
  stamp(s: Stamp): void {
    this.u64(s.hlc.wall);
    this.u16(s.hlc.logical);
    this.u64(s.replica);
  }
  value(v: Value): void {
    this.u8(VALUE_TAG[v.t]);
    switch (v.t) {
      case "null":
        break;
      case "bool":
        this.u8(v.v ? 1 : 0);
        break;
      case "i64":
        this.i64(v.v);
        break;
      case "f64":
        this.f64(v.v);
        break;
      case "str":
      case "frac":
        this.bytes(utf8.encode(v.v));
        break;
      case "color":
        this.u32(v.v);
        break;
      case "ref":
        this.u64(v.v.replica);
        this.u64(v.v.counter);
        break;
    }
  }
  finish(): Uint8Array {
    return this.buf.slice(0, this.pos);
  }
}

class Reader {
  private view: DataView;
  pos = 0;
  constructor(
    private readonly buf: Uint8Array,
    start: number,
  ) {
    this.view = new DataView(buf.buffer, buf.byteOffset, buf.byteLength);
    this.pos = start;
  }
  private need(n: number): void {
    if (this.pos + n > this.buf.length) throw new Error("unexpected end of snapshot");
  }
  u8(): number {
    this.need(1);
    return this.view.getUint8(this.pos++);
  }
  u16(): number {
    this.need(2);
    const v = this.view.getUint16(this.pos, true);
    this.pos += 2;
    return v;
  }
  u32(): number {
    this.need(4);
    const v = this.view.getUint32(this.pos, true);
    this.pos += 4;
    return v;
  }
  u64(): bigint {
    this.need(8);
    const v = this.view.getBigUint64(this.pos, true);
    this.pos += 8;
    return v;
  }
  i64(): bigint {
    this.need(8);
    const v = this.view.getBigInt64(this.pos, true);
    this.pos += 8;
    return v;
  }
  f64(): number {
    this.need(8);
    const v = this.view.getFloat64(this.pos, true);
    this.pos += 8;
    return v;
  }
  string(): string {
    const n = this.u32();
    this.need(n);
    const s = utf8d.decode(this.buf.subarray(this.pos, this.pos + n));
    this.pos += n;
    return s;
  }
  stamp(): Stamp {
    const wall = this.u64();
    const logical = this.u16();
    const replica = this.u64();
    return { hlc: { wall, logical } satisfies Hlc, replica };
  }
  value(): Value {
    const tag = this.u8();
    switch (tag) {
      case 0:
        return { t: "null" };
      case 1:
        return { t: "bool", v: this.u8() !== 0 };
      case 2:
        return { t: "i64", v: this.i64() };
      case 3:
        return { t: "f64", v: this.f64() };
      case 4:
        return { t: "str", v: this.string() };
      case 5:
        return { t: "color", v: this.u32() };
      case 6:
        return { t: "frac", v: this.string() };
      case 7: {
        const r = this.u64();
        const c = this.u64();
        return { t: "ref", v: opId(r, c) };
      }
      default:
        throw new Error(`bad value tag ${tag}`);
    }
  }
}

export function canonicalBytes(doc: Document): Uint8Array {
  const w = new Writer();
  const entries = doc.entries();
  w.u64(BigInt(entries.length));
  for (const [id, obj] of entries) {
    w.u64(id.replica);
    w.u64(id.counter);
    w.u8(obj.kind === null ? 0 : KIND_TAG[obj.kind]);
    w.u8(obj.deleted.value ? 1 : 0);
    w.stamp(obj.deleted.at);
    const keys = [...obj.props.keys()].sort(); // ASCII keys: code-unit order == byte order
    w.u32(keys.length);
    for (const k of keys) {
      const reg = obj.props.get(k)!;
      w.bytes(utf8.encode(k));
      w.value(reg.value);
      w.stamp(reg.at);
    }
  }
  return w.finish();
}

export function hash(doc: Document): Uint8Array {
  return blake3(canonicalBytes(doc));
}

export function hex(bytes: Uint8Array): string {
  let s = "";
  for (const b of bytes) s += b.toString(16).padStart(2, "0");
  return s;
}

export function unhex(s: string): Uint8Array {
  const out = new Uint8Array(s.length / 2);
  for (let i = 0; i < out.length; i++) out[i] = parseInt(s.slice(i * 2, i * 2 + 2), 16);
  return out;
}

export function hashHex(doc: Document): string {
  return hex(hash(doc));
}

export function encodeSnapshot(doc: Document): Uint8Array {
  const body = canonicalBytes(doc);
  const out = new Uint8Array(5 + body.length);
  out.set(SNAPSHOT_MAGIC, 0);
  out[4] = SNAPSHOT_VERSION;
  out.set(body, 5);
  return out;
}

export function decodeSnapshot(bytes: Uint8Array): Document {
  if (
    bytes.length < 5 ||
    SNAPSHOT_MAGIC.some((b, i) => bytes[i] !== b) ||
    bytes[4] !== SNAPSHOT_VERSION
  ) {
    throw new Error("bad snapshot header");
  }
  const r = new Reader(bytes, 5);
  const doc = new Document();
  const n = r.u64();
  for (let i = 0n; i < n; i++) {
    const replica = r.u64();
    const counter = r.u64();
    const kindTag = r.u8();
    const state = new ObjectState();
    if (kindTag !== 0) {
      const kind = TAG_KIND[kindTag];
      if (!kind) throw new Error(`bad kind tag ${kindTag}`);
      state.kind = kind;
    }
    const deletedV = r.u8() !== 0;
    const deletedAt = r.stamp();
    state.deleted = { value: deletedV, at: deletedAt };
    const np = r.u32();
    for (let p = 0; p < np; p++) {
      const key = r.string();
      const value = r.value();
      const at = r.stamp();
      const reg: Register<Value> = { value, at };
      state.props.set(key, reg);
    }
    doc.insertRaw({ replica, counter } satisfies OpId, state);
  }
  if (r.pos !== bytes.length) throw new Error("trailing bytes in snapshot");
  return doc;
}
